lms/static/js/discovery/app.js

```diff
@@ -1,69 +1,111 @@
-            filters.changeQueryFilter(query);
+            search.performSearch(query, filters.getTerms());
         });
 
-        dispatcher.listenTo(filters, 'search', function (searchTerm, facets) {
-            collection.performSearch(searchTerm, facets);
+        dispatcher.listenTo(refineSidebar, 'selectOption', function (type, query, name) {
             form.showLoadingIndicator();
+            if (filters.get(type)) {
+                removeFilter(type);
+            }
+            else {
+                filters.add({type: type, query: query, name: name});
+                search.refineSearch(filters.getTerms());
+            }
         });
 
-        dispatcher.listenTo(filters, 'clear', function () {
-            form.clearSearch();
-            collection.performSearch();
-            filters.hideClearAllButton();
+        dispatcher.listenTo(filterBar, 'clearFilter', removeFilter);
+
+        dispatcher.listenTo(filterBar, 'clearAll', function () {
+            form.doSearch('');
         });
 
-        dispatcher.listenTo(results, 'next', function () {
-            collection.loadNextPage();
-            form.showLoadingIndicator();
+        dispatcher.listenTo(listing, 'next', function () {
+            search.loadNextPage()
         });
 
-        dispatcher.listenTo(collection, 'search', function () {
-            if (collection.length > 0) {
-                form.showFoundMessage(collection.totalCount);
-                results.render();
+        dispatcher.listenTo(search, 'next', function () {
+            listing.renderNext();
+        });
+
+        dispatcher.listenTo(search, 'search', function (query, total) {
+            if (total > 0) {
+                form.showFoundMessage(total);
+                if (query) {
+                    filters.add(
+                        {type: 'search_query', query: query, name: quote(query)},
+                        {merge: true}
+                    );
+                }
             }
             else {
-                form.showNotFoundMessage(collection.searchTerm);
+                form.showNotFoundMessage(query);
+                filters.reset();
             }
-            facetsBarView.renderFacets(collection.facets);
-            form.hideLoadingIndicator();
-        });
-
-        dispatcher.listenTo(collection, 'next', function () {
-            results.renderNext();
             form.hideLoadingIndicator();
+            listing.render();
+            refineSidebar.render();
         });
 
-        dispatcher.listenTo(collection, 'error', function () {
+        dispatcher.listenTo(search, 'error', function () {
             form.showErrorMessage();
             form.hideLoadingIndicator();
         });
 
-        dispatcher.listenTo(facetsBarView, 'addFilter', function (data) {
-            filters.addFilter(data);
-        });
-
         // kick off search on page refresh
         form.doSearch(searchQuery);
 
+        function removeFilter(type) {
```

Comment:
Why not dispatch removeFilter event to all views and let them handle it as they wish, instead of writing this function in app.js? Doesn't seem to belong here.